tests/test_cli.py

```diff
@@ -101,69 +105,165 @@ def test_validate_missing_required_args(self):
-            assert f"Results saved to {output_path}" in result.output
+            for uuid in expected_skipped_uuids:
+                assert f"AIP '{uuid}' already validated, skipping." in result.output
 
-            output_data = pd.read_csv(output_path)
-            assert len(output_data) == 1
-            assert bool(output_data["valid"].iloc[0]) is True
+    def test_bulk_validate_incremental_writes_during_thread_failures(
+        self,
+        tmp_path,
+    ):
+        """Test that even when individual validation threads fail, content is still
+        getting written to output CSV."""
+        output_csv = str(tmp_path / "output.csv")
 
-    def test_bulk_validate_invalid_csv(self):
         runner = CliRunner()
-        result = runner.invoke(
-            cli,
-            [
+        with patch("lambdas.cli.validate_aip_via_lambda") as mock_validate:
+            mock_validate.side_effect = [
+                {
+                    "valid": True,
+                    "elapsed": 1.5,
+                    "s3_uri": "s3://bucket/test",
+                },
+                {
+                    "valid": True,
+                    "elapsed": 1.5,
+                    "s3_uri": "s3://bucket/test",
+                },
+                SystemExit(),  # mocks an AIP validation thread that quits unexpectedly
+                SystemExit(),  # mocks an AIP validation thread that quits unexpectedly
+            ]
+
+            args = [
+                "--verbose",
                 "bulk-validate",
                 "-i",
-                "tests/fixtures/cli/bulk-validation/invalid_input.csv",
-            ],
-        )
-        assert result.exit_code == 1
-        assert (
-            "Input CSV must have 'aip_uuid' and/or 'aip_s3_uri' columns" in result.output
+                "tests/fixtures/cli/bulk-validation/input_with_existing_results.csv",
+                "-o",
+                output_csv,
+            ]
+
+            runner.invoke(cli, args)
+
+        # two rows written despite other threads failing
+        output_df = pd.read_csv(output_csv)
+        assert len(output_df)
+
+    def test_bulk_validate_existing_incrementally_writes_during_parallel_validation(
+        self, tmp_path, mocker, reraise
+    ):
+        """Test that as the threaded worker method validate_aip_bulk_worker runs, rows
+        are getting written to the output CSV.  This is important: if the CLI process
+        quits unexpectedly, we need to know that each threaded worker had already
+        written its content."""
+        output_csv = str(tmp_path / "output.csv")
+
+        original_worker = cli_module.validate_aip_bulk_worker
+
+        def wrapper_worker(*args, **kwargs):
+            result = original_worker(*args, **kwargs)
+
+            # this use of 'reraise' (from 'pytest-reraise' library) is required for
+            # bubbling up assertion failures that occur as part of a multithreaded process
+            with reraise:
```

Comment:
This is the application of `pytest-reraise` that allows `assert` statements that are acting on threaded calls to still bubble up and fail the test if they _do_ fail.

Reply:
I thought so too!  I couldn't figure out why intentionally failed assertions inside that block were still passing tests.  This library was just the needed solution.  There _are_ ways to do this yourself, but clunky and error prone.